### PR TITLE
backupccl: reduce over-splitting during RESTORE

### PR DESCRIPTION
**Pull request by dt**
When constructing the spans to restore, we can produce fewer, wider spans
that each contain more files to avoid splitting as much.

Previously we split at the bounds of every file in the base backup, then
added incremental layer files to any span they overlapped. This changes
that slightly to keep appending files that would create a new span at the
right-hand-side of the span covering to instead extend current rightmost
entry in the covering if the total size of what has been added to it is
still below our target size per span (384mb).

Release justification: bug fix.

Release note (bug fix): reduce the amount that RESTORE over-splits.

### CODE REVIEW

**Comment by dt** [comment]
Here's a of the same backup we use in the restore2tb backup -- with is roughly 660GB data size -- being restored with and without (an earlier version of) this patch. Note the axis labels:

Screen Shot 2022-08-19 at 6 34 11 PM

Screen Shot 2022-08-19 at 6 34 26 PM

I think I was using a slightly different target size and lower guess sizes for that run (I believe they were 448 and 8mb, but I could be wrong). I did notice some rebalancing traffic on the merged spans cluster however, indicating we were not splitting quite enough to avoid any post-ingest rebalance:
Screen Shot 2022-08-19 at 6 46 58 PM
I suspect we'd benefit from enabling as-we-fill splits in the sst batcher to avoid these post-ingest rebalances. The amount however isn't very concerning and didn't have a noticeable impact on top-line restore time:
41:00 on the control master, and 40:17 on the experimental merging build. The merging build's rebalancing did mean that final disk usage on that cluster was 2.1% higher than the master one where everything went straight to L6 and never moved.

**Comment by msbutler** [comment]
wow, beat me to it! I've got another impl that seems slightly more involved [here](https://github.com/msbutler/cockroach/commit/1bf0004645e6e245b388e59c6d4cddf72183480a). I'll plan to dig a little deeper here next week.

**Comment by msbutler** [comment]
will also consider how this jives with @itsbilal [pr](https://github.com/cockroachdb/pebble/pull/1905)

**Comment by dt** [comment]
@msbutler took a peek at your version, but I _think_ prefer just widening the rightmost cover span as we go vs needing to mess with the the covering after. I think I'd suggest you take this PR over instead but up to to. I think the biggest thing I'd want to do here to merge this is add another test case where the size isn't zero, and maybe make it a cluster setting. I'd also like to enable as-we-go-splits in the batcher but I think my roachtest run shows that that isn't a strict requirement.

On the level merging iter: I'm not really sure this changes things? This is putting more `Files` into each span, but there was already no real guarantees about sstables -- not files -- not overlapping in a level (indeed they overlap all the time). The only difference is previously in the _first level_ we'd never assign multiple Files to the same span, but now we will, but we can and would often assign multiple files to the same Span in inc levels, so there was already the potential for overlapping sstables (since even thought the files are non-overlapping, they can be in different sstables which _are_ overlapping), right?

**Comment by msbutler** [comment]
ah rats, Bilal's PR cannot get applied to RESTORE. I was conflating `files` and `SSTable` in `makeSimpleImportSpans`

**Comment by msbutler** [comment]
The high level difference between David's implementation and mine is:
- my implementation expands the spans `after` the `restoreSpanEntry` has formed, which provides an upper bound on the amount of work in a span. The upper bound prevents mid/post RESTORE range splits (full of data).
- David's expands the spans _during_ `restoreSpanEntry` formation and only when evaluating the 0th layer (the full backup spans). This impl cannot guarantee an upper bound of work on a given span. So, if incremental backups add significantly more work to a span, this implementation could cause mid/post restore range splits, which are _much_ more expensive than the mid/post restore range merges that we're currently dealing with. 

Given all this, I think it makes sense to go with david's lighter weight implementation, as long as we also implement range splitting in the batcher.

**Comment by msbutler** [comment]
@dt per weekly standup, I'm handing this back to you.

**Comment by dt** [comment]
Added another commit to enable as-we-fill splits of full ranges, replacing the disabling of batcher's splits during restore more narrow, to be just a disabling of scatters.
Results look promising (note the axis labels i.e the magnitude of these are not comparable):

master, all pre-split and very overspilt (35min, 21602 ranges)
Screen Shot 2022-08-30 at 9 11 33 PM

just bigger spans, no as-we-go splits (32min, 1976 ranges):
Screen Shot 2022-08-30 at 9 11 15 PM

bigger spans plus enabling as-we-go-splits (29min, 1953 ranges):
Screen Shot 2022-08-30 at 9 11 48 PM

**Comment by dt** [comment]
> how late we are in the release, I don't think any of this should be enabled by default.

I'm tempted to enable it, and potentially even backport too, to fix the bad behavior of over-splitting, at least to 22.1.

**Comment by shermanCRL** [comment]
> > how late we are in the release, I don't think any of this should be enabled by default.

> I'm tempted to enable it, and potentially even backport too, to fix the bad behavior of over-splitting, at least to 22.1. I'm not _that_ scared of it, since we're just changing how we divide the spans, not what we do with them or how we do it really.

You can make the call, I’m ok with it if the tests are good (including the cluster testing we plan to do).

**Review by msbutler — COMMENTED** [review]
I have two high level worries about this diff (that are perhaps off base): 
First, this diff makes it easier for one node to end up processing significantly more data, relative to other nodes in the flow. This has a couple risky consequences:
1. the restore runtime can get bottlenecked by one node
2. kv may feel compelled to move full ranges from the hot node, which is probably more expensive than merging small ranges

Here's one way this could happen, even with range splitting enabled:
- suppose the incremental backups add significantly more data to a single `restoreSpanEntry` that `makeSimpleImportSpans` has grown, as the full backup contained relatively less data in this span. Also suppose that before this diff, the data from the incremental backups would have been divided into several `restoreSpanEntry`'s that also got divided over several import span chunks.
- before this diff, the incremental backup data was divided across chunks, and therefore, work was distributed across the cluster.
- with this diff, it is more likely that al the incremental backup data will end up in fewer chunks, and therefore get processed by fewer nodes. 

Second, given that this diff does not change our checkpointing procedure (something we said should happen after this work), the larger spans this work creates makes checkpointing more imprecise, which could make restore job pausing/resuming more expensive. (i.e. https://github.com/cockroachdb/cockroach/issues/81116, though disallowShadowing [may have been disabled](https://github.com/cockroachdb/cockroach/issues/81116#issuecomment-1135128433)? ).

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (dt) [reply pkg/ccl/backupccl/restore_span_covering.go:102]: old backups don't set EntryCounts.

**Comment by dt** [comment]
> suppose the incremental backups add significantly more data to a single restoreSpanEntry that makeSimpleImportSpans has grown, as the full backup contained relatively less data in this span

I thought a little about this initially, but came away thinking this is true _anyway_, even without merging base spans at all, so we shouldn't let it stop us from fixing the issue at hand. If we think about when we might hit this case -- where the base backup has very little data in some span but then later layers have much more in that span, that is going to be most common when you create a table, back it up, then load data into it, i.e. many ranges of data, then backing up incrementally; all those big, full ranges of the inc backup all fall within the one span in the inc backup. This however is all true without merging -- it's a single span per table when it is empty -- so the only difference with merging is magnitude, but my thesis here is that magnitudes that are unfortunate are reachable anyway.

If we decide that we care about this -- I'm not sure we really do right now -- then I think we'd want to do something more general, that post-processes the covering if any one chunk is more than X% (probably 1/nodes*2 or something), but I don't really see it as coupled to merging small base spans.

> does not change our checkpointing procedure (something we said should happen after this work)

I don't think we should make extra spans here just for the sake of progress checkpoints since we already know that that does not actually make progress lose that much less; our losses are already due to out-of-order span completion, not individual span sizes, and even post-merging our spans sizes here are still, in absolute numbers, generally small. I don't really find the progress effects that compelling: this was always intended to have ~range sized chunks, is currently dramatically undershooting that, and 8mb or 64mb or 364mb or even 3gb are all tiny numbers in TB restores, where we really care about progress. In a tiny restore where the difference here is actually a real % of the total size, then the total size is small enough to not really care about progress much anyway since it is small anyway (also there's a minimum time between progress updates too).

**Comment by msbutler** [comment]
Cool, I'm no longer worried about merging creating an uneven distribution of work.

Sorry, I should have been more clear on why I worry about checkpointing. Like you, I don't care that much about the additional lost work, given that the spans are larger. I worry that the larger spans will cause Restore to slow to crawl more frequently due to CheckForSSTCollisions. In [22.1](https://github.com/cockroachdb/cockroach/pull/82357) and 21.2, we set `disallowShadowing` to false. Perhaps in this PR or after the branch cut, we should set disallowShadowing to false in 22.2 as well.

**Comment by msbutler** [comment]
> The lost-work and the disallowShadowing shadowing overlaps are one and the same though, aren't they?
Yes, both are caused by our naive ordering-sensitive checkpointing strategy, which, as you say, causes us to reingest a bunch of chunks again. My point is: if the avg span entry size is significantly bigger and the number of span entries retried remains constant (not sure if this is correct), then this diff means we're more susceptible to the check for key collisions problem. Regardless, I think we should set `disallowShadowing` == false for 22.2, as we have not yet updated our checkpointing scheme, which means my point is moot.

File: pkg/ccl/backupccl/restore_span_covering_test.go
Comment (dt) [reply pkg/ccl/backupccl/restore_span_covering_test.go:127]: Yep, done!

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (dt) [reply pkg/ccl/backupccl/restore_span_covering.go:32]: done

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (msbutler) [inline pkg/ccl/backupccl/restore_span_covering.go:102]: maybe add def `curGovGrowSize`? I think it's: curCovGrowSize is the number of bytes added to the right most cover since its creation.

File: pkg/ccl/backupccl/restore_span_covering.go
Comment (msbutler) [inline pkg/ccl/backupccl/restore_span_covering.go:148]: I might be missing something here. why do we need grow `curCovGrowSize` here for when `i==l`? 

I thought this loop is growing "older" covers (i.e. not the current, rightmost cover), potentially beyond the targetSize limit.  I'm under the impression that curCovGrowSize tracks the right most cover's size, so I don't follow why it gets incremented here.

Perhaps this logic could use another comment. It's quite hard to follow all the indexing!

Comment (dt) [reply pkg/ccl/backupccl/restore_span_covering.go:148]: `i == l` is the rightmost

Comment (msbutler) [reply pkg/ccl/backupccl/restore_span_covering.go:148]: didn't we already grow the right most span above?

Comment (msbutler) [reply pkg/ccl/backupccl/restore_span_covering.go]: aha! This makes sense now. thanks for clarifying!

**Comment by msbutler** [comment]
The non test code looks good to me. Last question: how confident are you that the test cases you put in exercise every nook and cranny of  the new makeSimpleImport spans? Before backporting, I wonder if we ought to write a few more specific test cases to ensure this. `makeSimpleImportSpans` is no longer so simple :)

**Review by msbutler — APPROVED** [review]

**Comment by dt** [comment]
> how confident are you that the test cases you put in exercise every nook and cranny of the new makeSimpleImport spans

The random test does a good job of ensuring correctness (it caught a bug here). It doesn't do much in terms of proving we merged into any specific number of spans for perf, but it proves we have the right files in the right spans I think, which makes me feel better about this.

**Comment by dt** [comment]
TFTR!

bors r+

**Comment by dt** [comment]
I tagged it for backport; we can see how long we want to bake on 22.2 before we merge it

**Comment by dankinder** [comment]
@dt I have been waiting eagerly for this change as I'm still blocked on restoring, per https://github.com/cockroachdb/cockroach/issues/86470 so thanks for cranking it out.

I'm currently on V22.1.5. I'll either need to wait until this gets backported to 22.1, or deploy the next alpha of 22.2, or build it myself. If you have any suggestion on which is best or the timeline of when these will happen I'd appreciate it.

**Pull request merged**
